examples/gnrc_border_router: Add ethos and start_network.sh to README.md

kYc0o opened the pull request:
I had added it in a wrong branch, open again with correct commit message
Added documentation for ethos and automatic setup with the start_network.sh script on #4725

authmillenon commented on examples/gnrc_border_router/README.md:
```diff
@@ -60,3 +60,98 @@ for further help.
+
+
+# gnrc_border_router with ethos
+A new setup is possible thanks to \#4438(https://github.com/RIOT-OS/RIOT/pull/4438) on which only one serial interface is needed. This is done through the ethos (Ethernet Over Serial) driver that multiplexes serial data between ethernet packets and shell commands.
```
It's not typical to reference a merged PR in a documentation text.

authmillenon replied:
Also, please try to keep the lines under 80 characters long.

kYc0o replied:
Yes we'll wait until the PR is merged, then I'll change that part

authmillenon replied:
But #4438 is already merged

kYc0o replied:
we are waiting for #4725

kYc0o replied:
You mean I should remove the reference to the closed PR?

authmillenon replied:
Yes. :-)

authmillenon commented on examples/gnrc_border_router/README.md:
````diff
@@ -60,3 +60,98 @@ for further help.
+```
+
+## Usage
+Once you flashed the BR example on your board, you should intialize your IP settings by choosing a prefix for your 6lo network. Just type `make term` and setup your BR:
````
Shouldn't you rather use the `ethos` tool for that?

authmillenon replied:
Ah the tun isn't initialized yet... Got it. (but maybe write below to kill this `make term` to prevent multiplexing issues over multiple terminals)

kYc0o replied:
Oh yes! sorry, I'll change that

authmillenon commented on examples/gnrc_border_router/README.md:
````diff
@@ -60,3 +60,98 @@ for further help.
+```
+
+At this point you should have these two programs compiled, necessary to compile your gnrc_border_router ** DO NOT COMPILE IT BEFORE ethos AND uhcp **.
+Then, proceed to compile and flash gnrc_border_router to your board as explained before.
````
Oops the formatting is a little broken here.

kYc0o replied:
In the web editor looked well, I don't know why in the diff it is formatter that way...

authmillenon replied:
Because it gets confused about the `** <text><linebreak><text> **` above

authmillenon replied:
(just remove the spaces inside the boldness stars)

kYc0o replied:
Ok thanks

gebart commented:
Can you change the prefix to anything in the ULA range (fc00::/7) instead of affe:: to use the RFC 4193 designated private use addresses

authmillenon commented:
`2001:0000::/29 - 2001:01F8::/29` are also [viable for experimental use](https://tools.ietf.org/html/rfc4727#section-3.4.1)

authmillenon commented on examples/gnrc_border_router/README.md:
```diff
@@ -60,3 +60,99 @@ for further help.
+
+
+# gnrc_border_router with ethos
+A new setup is possible on which only one serial interface is needed. This is done through the ethos (Ethernet Over Serial) driver that multiplexes serial data between ethernet packets and shell commands.
```
``s/ethos/`\0`/g``

authmillenon replied:
And still too long lines

kYc0o replied:
Why the long lines bother you? Do you think I should split into 2 paragraphs?

authmillenon commented on examples/gnrc_border_router/README.md:
```diff
@@ -60,3 +60,99 @@ for further help.
+A new setup is possible on which only one serial interface is needed. This is done through the ethos (Ethernet Over Serial) driver that multiplexes serial data between ethernet packets and shell commands.
+
+## Requirements
+This functionality works only in Linux machines, Mac support will be added in the future (lack of `tap` interface), and Windows support should possible by setting up the needed `tap` interface with an external program (not tested).
```
`s/Mac/Mac OSX/`?

authmillenon commented on examples/gnrc_border_router/README.md:
````diff
@@ -60,3 +60,99 @@ for further help.
+## Setup
+Take care about the modification of the `Makefile`, since by default SLIP is enabled.
+In order to select ethos as the driver, you should modify some lines as following:
+```make
````
For better readability in `vim` or `cat` I would prefer a newline between prose text and code blocks.

authmillenon commented on examples/gnrc_border_router/README.md:
````diff
@@ -60,3 +60,99 @@ for further help.
+```make
+#CFLAGS += -DSLIP_UART=$(SLIP_UART)
+#CFLAGS += -DSLIP_BAUDRATE=$(SLIP_BAUDRATE)
+CFLAGS += '-DETHOS_UART=UART_DEV(0)' -DETHOS_BAUDRATE=115200 -DUSE_ETHOS_FOR_STDIO
````
I'm wondering if a diff wouldn't be better understandable, but I'm the git pro here :P

kYc0o replied:
Hahaha yeah, but you can teach me how to do it and we can try ;-)

authmillenon commented on examples/gnrc_border_router/README.md:
````diff
@@ -60,3 +60,99 @@ for further help.
+This will allow to compile the BR firmware using ethos.
+Thus:
+```bash
+$ make clean all flash
````
`s/^$ //` if you syntax highlight as bash (otherwise `$` will be parsed as variable I think)

kYc0o replied:
I didn't really catch `s/^$ //` ... :-(

OlegHahm replied:
@authmillenon proposed to strip the `$` from the line.

gebart commented:
@authmillenon I interpret that paragraph that they mean more for routing experiments across sites for developing the IPv6 internet, not for local IoT experiments.

authmillenon commented on examples/gnrc_border_router/README.md:
```diff
@@ -60,3 +60,99 @@ for further help.
+Now, you should be able to ping your nodes using the global addres starting by your prefix, on our case `affe::`.
+
+# gnrc_border_router using automatic configuration (not merged yet)
+In \#4725 (https://github.com/RIOT-OS/RIOT/pull/4725) an effort to automate all the previous steps is done using 3 tools:
```
You can have proper links in markdown doing this

```markdown
[\#4725](https://github.com/RIOT-OS/RIOT/pull/4725)
```

(and it's not too disruptive for local readers either)

authmillenon commented:
> @authmillenon I interpret that paragraph that they mean more for routing experiments across sites for developing the IPv6 internet, not for local IoT experiments.

Well the RFC also references `2001:db8::/32` for [documentation purposes](https://tools.ietf.org/html/rfc3849#section-2) which is exactly what we have here. :-)

bapclenet commented:
Ping @authmillenon

kYc0o commented:
Oops! That's me, I'll address the comments ASAP

kYc0o commented:
I addressed most of the comments and changed the addresses for "documentation purposes" to `2001:db8::/32`.

emmanuelsearch commented:
@kYc0o what's the status on this PR, on your side?
@aabadie do you see ways to improve this README based on your recent experience?

kaspar030 commented:
Current #4725 drops the slip mode completely. Should the readme reflect this?

OlegHahm commented:
> Current #4725 drops the slip mode completely. Should the readme reflect this?

I think it might be nice to add a hint that you can use SLIP (or any Ethernet transceiver) as well.

kYc0o commented:
On my side I think I addressed most of the comments, maybe a paragraph is apparently too large but I can't see how to split it or male it short. Otherwise we are still waiting for #4725

kYc0o commented:
Looking at #4725 there are lots of changes so it seems this is very outdated. @kaspar030 can you confirm what is up-to-date and what's not?

kYc0o commented:
As for my last tests, this guide is still valid, aiming to merge it once #4725 is merged too?

kaspar030 commented:
ACK after you shorten the lines to <= 80 characters. :)

kYc0o commented:
Oh yes, I'm about to re-organise it because SLIP is not present anymore, I'll put first the config with the script and then with ethos, in case of someone wants to try the manual config with it. Some rephrase is needed too (#4725 is merged so we don't need to merge it locally anymore, etc.), but I think every step is here.

kYc0o commented:
BTW, we'll get rid of SLIP?

kaspar030 commented:
> we'll get rid of SLIP?

I don't think so, as it's still perfectly working and standards compliant.

aabadie commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,144 @@
+A new setup is possible on which only one serial interface is needed. 
+This is done through the ethos (Ethernet Over Serial) driver and UHCP.
+Ethos multiplexes serial data to separate ethernet packets from shell commands.
+UHCP is in charge of configure the wireless interface prefix and routes on the BR.
```
You should describe what UHCP means here (you did that a few lines below). Also `configure` => `configuring`

aabadie commented on examples/gnrc_border_router/README.md:
````diff
@@ -1,4 +1,144 @@
+```bash
+sudo sh start_network.sh /dev/ttyACMx tap0 2001:db8::/64
+```
+
````
I think it would be great to mention here that one has access to the shell here. For instance, one can track the neighbour cache using `ncache`.

aabadie commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,144 @@
+This prefix should be announced to other motes through the wireless interface.
+
+At this point you should be able to ping motes using the global address.
+You can do this through the RIOT console, for instance using `gnrc_networking`:
```
To me, this is not clear. do you want to ping the motes behind the BR ? so why talking about gnrc_networking ? It's confusing.

kYc0o replied:
I just want to point out that you can use global addresses to ping motes, either from linux or from other motes. I'll rephrase this to be more clear.

aabadie commented on examples/gnrc_border_router/README.md:
````diff
@@ -1,4 +1,144 @@
+```
+> ifconfig 6 add 2001:db8::2/48
+> ifconfig 5 add 2001:db8::3/64
+> fibroute add :: via <link-local of tap> dev 6
````
You should tell what type of interface are 5 and 6 (wired/wireless). Looking at haw the BR is configured by `start_network.sh`, the BR wired interface has a link-local address `fe80:`

aabadie commented on examples/gnrc_border_router/README.md:
````diff
@@ -33,21 +195,21 @@ the program by calling something like:
+ncache add 6 2001:db8::1
 ```
 After this you're basically done and should be able to ping between the border
 router and the outside world (assuming that the Linux host is properly
````
To me there is a step missing here (and some more detail as well), because at this point the prefix is not propagated to the motes behind the BR (that was the case for me):
* detail that interface `6` is the wired interface and `5` is the wireless one, just for clarity
* configure the wireless interface with a global address using the same prefix:
```
ifconfig 5 add 2001:db8::3
```

haukepetersen commented:
@aabadie: thanks for you input!

@kYc0o: could you adjust the README ASAP!? Would be good to merge this as soon as possible, as the border router app was merged and has a broken doc currently...

emmanuelsearch commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,144 @@
-# gnrc_networking_border_router example
+# gnrc_border_router using automatic configuration
+A new setup is possible on which only one serial interface is needed. 
```
Since this is now the default config, how about rephrasing the first two sentences with something like "This setup uses a single serial interface, ethos (Ethernet Over Serial) and UHCP (micro Host Control Protocol).

kaspar030 replied:
> UHCP (micro Host Control Protocol)


micro host configuration protocol ;)

emmanuelsearch commented on examples/gnrc_border_router/README.md:
````diff
@@ -1,4 +1,144 @@
+endif
+# include UHCP client
+USEMODULE += gnrc_uhcpc
+```
````
Actually, do we need this in the readme at this point?
It seems that this check is superfluous in this example at least.
Should we move this to another part of the readme, e.g. in the manual config?

emmanuelsearch commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,144 @@
+## Issues
+A common issue is that after ~15min global addresses on the BR Neighbour Cache dissappear.
+We are investigating why it happens and we'll propose a solution soon.
+If you experience it, just reboot your nodes and they will come again on-line.
```
reference the open issue on this here. I think this is #5122 right?

kYc0o commented:
Comments addressed. Sorry for the delay, I was sick this morning... Please check if all comments were correctly addressed.

emmanuelsearch commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,155 @@
+By typing `help` you will get the list of available shell commands.
+
+At this point you should be able to ping motes using the global address.
+For instance, if you use the `gnrc_networking` example, you can ping with:
```
"example" => "example on the mote"
"ping" => "ping it from your machine"

emmanuelsearch commented on examples/gnrc_border_router/README.md:
````diff
@@ -1,4 +1,155 @@
+```
+
+Just replace this address by your mote's address.
+Using `ifconfig` shows you the addresses of your mote.
````
Using `ifconfig`=> Using `ifconfig` on the shell of your mote

emmanuelsearch commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,155 @@
+Using `ifconfig` shows you the addresses of your mote.
+
+The script also sets up a link-local address on your Linux network interface.
+It can be pinged from 6lo motes:
```
link-local => ULA (Unique Local Address)
(quite different, because routable inside the administrative domain)

emmanuelsearch replied:
Maybe also add: check your ULA on your PC with `ifconfig` command

emmanuelsearch commented on examples/gnrc_border_router/README.md:
````diff
@@ -1,4 +1,155 @@
+> ping6 fd00:dead:beef::1
+```
+
+Thus far, you can communicate using 6lowPAN and IPv6 between motes and your PC.
````
Thus, IPv6 communication with between your PC and your motes is enabled?

emmanuelsearch commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,155 @@
+Thus far, you can communicate using 6lowPAN and IPv6 between motes and your PC.
+
+## Issues
+A common issue [#5122](https://github.com/RIOT-OS/RIOT/issues/5122) 
```
A known issue?

emmanuelsearch commented:
Thanks. Added a few comments. ACK when addressed.

kYc0o commented:
Thanks to you @emmanuelsearch.

aabadie commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,159 @@
+This is done through the same serial interface.
+By typing `help` you will get the list of available shell commands.
+
+At this point you should be able to ping motes using the global address.
```
using the global address => using their global address

aabadie commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,159 @@
+By typing `help` you will get the list of available shell commands.
+
+At this point you should be able to ping motes using the global address.
+For instance, if you use the `gnrc_networking` example on the mote, you can 
```
Maybe use a link to point to [gnrc_networking](https://github.com/RIOT-OS/RIOT/tree/master/examples/gnrc_networking) and give a real example: run ifconfig one mote and display the output highlighting the global address. Then the next command will be more explicit.

kYc0o replied:
> Maybe use a link to point to gnrc_networking and give a real example: run ifconfig one mote and display the output highlighting the global address.

Mmm cannot get the point here... What do you mean with a link to point to gnrc_networking?

aabadie replied:
I just meant to use the GitHub syntax to create an hyperlink to the gnrc_networking page:
```
[gnrc_networking](https://github.com/RIOT-OS/RIOT/tree/master/examples/gnrc_networking)
```

kYc0o replied:
Mmm ok I can do that

aabadie commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,159 @@
+mote.
+
+The script also sets up a ULA (Unique Local Address) address on your Linux 
+network interface.
```
Be explicit on the interface (I can have 4 interfaces on my Linux for instance).

aabadie commented on examples/gnrc_border_router/README.md:
```diff
@@ -1,4 +1,159 @@
+We are investigating why it happens and we'll propose a solution soon.
+If you experience it, just reboot your nodes and they will come again on-line.
+
+# gnrc_border_router with manual config
```
"Manual configuration of gnrc_border_router" ?

kYc0o replied:
I'd name it as it was originally, something like gnrc_border_router example with manual config, what do you think?

emmanuelsearch replied:
I guess it's fine. However: let's merge that asap, and also put it in the release candidate branch so that people have the required documentation to test...

kYc0o replied:
OK I'll push my last addressed comments ASAP and try to put everything in the RC.

kYc0o commented:
Since there's already an ACK, I'll squash the commits and merge. Since it is a readme I don't think Murdock is required for this. I'll let some minutes before doing it, in case of someone has more comments.

emmanuelsearch commented:
re-ACK. Merge at will (and back port to release branch asap).

kYc0o commented:
Squashed and go!

authmillenon commented:
Needs backport.
